### DISCUSSION TITLE
Converting a repo into a co-located repo

### DISCUSSION THREAD

**Discussion by nrbray** [jj-vcs/jj#4945] [Q&A] [answered]
**Note: since this discussion, the [documentation](https://github.com/martinvonz/jj/blob/main/docs/git-compatibility.md#converting-a-repo-into-a-co-located-repo) has kindly been updated by Tim.**

May I ask for clues for how to recover my content lost as below?

Perhaps my faux-pas may help me to elaborate on the Git compatibility documentation?

## Description

First time use of  ```jj git init .``` in some new work.  Then I found my nix project depended on a .git dir which was missing, so I just did what it said here for [Converting a repo into a co-located repo](https://github.com/martinvonz/jj/blob/main/docs/git-compatibility.md#converting-a-repo-into-a-co-located-repo).

Sadly I lost all content!

The jj log showed one previous entry and the jj op log showed many.

I could not for the life of me figure out how to restore my content.  Not with jj, nor with my marginally better understanding of git!

## Steps to Reproduce the Problem

```bash
# Move the Git repo
mv .jj/repo/store/git .git
# Tell jj where to find it
echo -n '../../../.git' > .jj/repo/store/git_target
# Ignore the .jj directory in Git
echo '/*' > .jj/.gitignore
# Make the Git repository non-bare and set HEAD
git config --unset core.bare
jj new @-  # edit: new documentation says do instead $ jj new && jj undo # that recovered files not checked in and fixed my issue here
```

## Expected Behavior

Some clues about how to create and preserve a git history in the  ```.jj/repo/store/git``` so if something goes wrong the repo can be recovered from the git contents.

## Actual Behavior

The moved hit repo contains no HEAD or other refspec.  The refs/jj/keep do not seem to be recognised by a git checkout.

## Specifications

```
nix run nixpkgs#neofetch
          ▗▄▄▄       ▗▄▄▄▄    ▄▄▄▖            nrb@dubedary 
          ▜███▙       ▜███▙  ▟███▛            ------------ 
           ▜███▙       ▜███▙▟███▛             OS: NixOS 24.05.20241023.32e940c (Uakari) x86_64 
            ▜███▙       ▜██████▛              Host: LENOVO 20JCS0LU06 
     ▟█████████████████▙ ▜████▛     ▟▙        Kernel: 6.6.58 
    ▟███████████████████▙ ▜███▙    ▟██▙       Uptime: 23 hours, 38 mins 
           ▄▄▄▄▖           ▜███▙  ▟███▛       Packages: 967 (nix-system), 570 (nix-user) 
          ▟███▛             ▜██▛ ▟███▛        Shell: bash 5.2.32 
         ▟███▛               ▜▛ ▟███▛         Resolution: 2160x1440 
▟███████████▛                  ▟██████████▙   DE: GNOME 46.2 (Wayland) 
▜██████████▛                  ▟███████████▛   WM: Mutter 
      ▟███▛ ▟▙               ▟███▛            WM Theme: Adwaita 
     ▟███▛ ▟██▙             ▟███▛             Theme: Adwaita [GTK2/3] 
    ▟███▛  ▜███▙           ▝▀▀▀▀              Icons: Adwaita [GTK2/3] 
    ▜██▛    ▜███▙ ▜██████████████████▛        Terminal: vscode 
     ▜▛     ▟████▙ ▜████████████████▛         CPU: Intel i7-7Y75 (4) @ 3.600GHz 
           ▟██████▙       ▜███▙               GPU: Intel HD Graphics 615 
          ▟███▛▜███▙       ▜███▙              Memory: 5580MiB / 15718MiB 
         ▟███▛  ▜███▙       ▜███▙
         ▝▀▀▀    ▀▀▀▀▘       ▀▀▀▘                                     
                                                                      

```

**Comment by tim-janik** [comment]
> First time use of `jj git init .` in some new work. Then I found my nix project depended on a .git dir which was missing, so I just did what it said here for [Converting a repo into a co-located repo](https://github.com/martinvonz/jj/blob/main/docs/git-compatibility.md#converting-a-repo-into-a-co-located-repo).

What did you do *after* `jj git init .` and *before* taking the below steps to move the repository?
Did you add files to the directory? Are those still there?
Did you run any `jj` commands after adding files? If so, jj should have created a snapshot commit.
Did you commit any files?

> Sadly I lost all content!

What is "content", files in the directory, commits as displayed by `jj log` or `git log`?

> The jj log showed one previous entry and the jj op log showed many.

Well, paste it here, at the least the first few entries and we might be able to give a hand.
Chances are, your files are in a commit and `jj new @-` just switched to an empty work dir, if you paste the log we can find the revision so `jj new <oldrevision>` will checkout your files again.

> I could not for the life of me figure out how to restore my content. Not with jj, nor with my marginally better understanding of git!
> ## Steps to Reproduce the Problem
> 
> ```shell
> # Move the Git repo
> mv .jj/repo/store/git .git
> # Tell jj where to find it
> echo -n '../../../.git' > .jj/repo/store/git_target
> # Ignore the .jj directory in Git
> echo '/*' > .jj/.gitignore
> # Make the Git repository non-bare and set HEAD
> git config --unset core.bare
> jj new @-
> ```

The command `jj new @-` will:
a) snapshot your files and store them in a commit,
b) then create a new working copy commit based on the *parent* that doesn't contain newly added files and make it the working copy.

See also: https://github.com/martinvonz/jj/pull/4948

**Reply by nrbray** [reply]
Thank you, and apologies for reporting a basic error on my part.  I'd like to understand what I did wrong, so have pasted the log below as it stands now after I've tried a few commands that I did not understand....  I know I should go read up, but was stabbing in the dark to see if I could get the files back.  It was only a few hours hacking where I thought I'd have a go at using jj - I should have taught myself jj separately first.


> What did you do _after_ `jj git init .` and _before_ taking the below steps to move the repository? Did you add files to the directory? Are those still there? Did you run any `jj` commands after adding files? If so, jj should have created a snapshot commit. Did you commit any files?

Files were in the directory before and added after the init.  I then used jj st, plus jj log, so was assuming that the files did not need a specific add.  I'd not made a commit - so perhaps that was my error, however I was just kicking myself for not having started with --colocate and wanted to get to that point before making any commit.


> What is "content", files in the directory, commits as displayed by `jj log` or `git log`?

By content, I mean the files in the directory, or any snapshot commit, if indeed anything had been put into the backing store.


> Well, paste it here, 

```sh
[nrb@dubedary:~/dir/WIP/learn/nix/hour59jj]$ jj log
◉  yppxsvtk?? jj-mtd@email.to1.uk 2024-11-21 20:26:45 d47b7c7b
│  (empty) (no description set)
│ ◉  yppxsvtk?? jj-mtd@email.to1.uk 2024-11-21 19:45:14 03d63dba
├─╯  (empty) (no description set)
│ @  qnvmqskk jj-mtd@email.to1.uk 2024-11-21 19:26:11 bca868ae
├─╯  (empty) (no description set)
◉  zzzzzzzz root() 00000000

[nrb@dubedary:~/dir/WIP/learn/nix/hour59jj]$ jj new qnv
Working copy now at: xuxnppov aac891e7 (empty) (no description set)
Parent commit      : qnvmqskk bca868ae (empty) (no description set)

[nrb@dubedary:~/dir/WIP/learn/nix/hour59jj]$ ll
total 0

[nrb@dubedary:~/dir/WIP/learn/nix/hour59jj]$ jj log
@  xuxnppov jj-mtd@email.to1.uk 2024-11-23 19:21:14 aac891e7
│  (empty) (no description set)
◉  qnvmqskk jj-mtd@email.to1.uk 2024-11-21 19:26:11 HEAD@git bca868ae
│  (empty) (no description set)
│ ◉  yppxsvtk?? jj-mtd@email.to1.uk 2024-11-21 20:26:45 d47b7c7b
├─╯  (empty) (no description set)
│ ◉  yppxsvtk?? jj-mtd@email.to1.uk 2024-11-21 19:45:14 03d63dba
├─╯  (empty) (no description set)
◉  zzzzzzzz root() 00000000

[nrb@dubedary:~/dir/WIP/learn/nix/hour59jj]$ jj op log
@  58f66ea09dff nrb@dubedary 28 seconds ago, lasted 78 milliseconds
│  new empty commit
│  args: jj new qnv
◉    46afcdbce3b4 nrb@dubedary 1 day ago, lasted 1 millisecond
├─╮  reconcile divergent operations
│ │  args: jj op log
◉ │    b15cace677c4 nrb@dubedary 1 day ago, lasted less than a microsecond
├───╮  reconcile divergent operations
│ │ │  args: jj op log
◉ │ │  9112403df640 nrb@dubedary 1 day ago, lasted 26 milliseconds
│ │ │  new empty commit
│ │ │  args: jj new @-
│ │ ◉  bfe38f0a42e3 nrb@dubedary 1 day ago, lasted 44 milliseconds
├───╯  restore into commit 832883ce972cfae99b0e1d8cfceb8f9af6e3b336
│ │    args: jj restore '--at-op=fec'
│ ◉  faf2d235a220 nrb@dubedary 1 day ago, lasted 44 milliseconds
├─╯  restore into commit 832883ce972cfae99b0e1d8cfceb8f9af6e3b336
│    args: jj restore '--at-op=fecf27d8b423'
◉  fecf27d8b423 nrb@dubedary 1 day ago, lasted 14 milliseconds
│  snapshot working copy
│  args: jj new @-
◉  9ae042b57ce5 nrb@dubedary 2 days ago, lasted 14 milliseconds
│  snapshot working copy
│  args: jj st
◉  cc7b6bddfc06 nrb@dubedary 2 days ago, lasted 15 milliseconds
│  snapshot working copy
│  args: jj st
◉  881d7dbb8d90 nrb@dubedary 2 days ago, lasted 60 milliseconds
│  add workspace 'default'
◉  000000000000 root()
```

**Reply by tim-janik** [reply, marked as answer]
Ok, looks like you created the repo (op 881d7dbb8d90), jj took a snapshot of your files (op cc7b6bddfc06), later it took a snapshot of your modifications (op 9ae042b57ce5) and then you started the co-location procedure. In operation fecf27d8b423, a new commit was created based on the parent commit of your working copy. Which means your files got stored away in the commit created in op 9ae042b57ce5, a new commit was created *without* the files and checked out.

Here are a few options you now have:

```sh
# Take close look at all operations including the diff of your files:
jj op log -p -s
```

```sh
# Take a look at the last commit that included your files, write down the commit ID:
jj log -p -s -T builtin_log_detailed  --at-op=9ae042b57ce5
# Create a new commit based on that commit id, this will checkout your files
jj new <commit-id>
```

Or, the following long command should do both steps for you (note the backticks to get the commit_id):
```sh
jj new  `jj log --no-graph -r @  -T commit_id --at-op=9ae042b57ce5`
```

**Reply by nrbray** [reply]
Wow, thank you.  That got my files back :)  I used your description below and learnt a lot.

Sorry for my delay.  Sounds made up, but I could not use those commands as I was on Jujutsu v0.17.1, the default for nixos-24.05.  My battery failed mid rebuild to nixos unstable, my SSD started failing, I moved machine, the graphics card went, and before I could complete I hit a cap of 50G data on my SIM.  I ran those commands with Jututsu v0.23.0

> ```shell
> # Take a look at the last commit that included your files, write down the commit ID:
> jj log -p -s -T builtin_log_detailed  --at-op=9ae042b57ce5
> # Create a new commit based on that commit id, this will checkout your files
> jj new <commit-id>
> ```

```txt
[nrb@lojijove:~/lojijove/hour59jj]$ jj new 1034710cf836e0d6ec67405ba250663ff59f0299
Working copy now at: snqtownp 09d81c7e (empty) (no description set)
Parent commit      : yppxsvtk?? 1034710c (no description set)
Added 9 files, modified 0 files, removed 0 files
```

**Comment by martinvonz** [comment]
It doesn't reproduce for me. Here's what I did:

```
at 08:38:43 ❯ cd (mktemp -d)

🌐 vonz in /tmp/tmp.CIO7uhL2M8
at 08:38:44 ❯ jj git init
Initialized repo in "."

🌐 vonz in /tmp/tmp.CIO7uhL2M8
at 08:38:53 ❯ ech^C

🌐 vonz in /tmp/tmp.CIO7uhL2M8
at 08:38:53 ❯ echo foo > file

🌐 vonz in /tmp/tmp.CIO7uhL2M8
at 08:38:58 ❯ jj log
@  u martinvonz@google.com 2024-11-22 08:41:36 8e3419c1
│  (no description set)
◆  z root() 00000000

🌐 vonz in /tmp/tmp.CIO7uhL2M8
at 08:41:36 ❯ jj commit -m foo
Working copy now at: x 2d33a4bf (empty) (no description set)
Parent commit      : u a2ae0e30 foo

🌐 vonz in /tmp/tmp.CIO7uhL2M8
at 08:41:43 ❯ jj log
@  x martinvonz@google.com 2024-11-22 08:41:42 2d33a4bf
│  (empty) (no description set)
○  u martinvonz@google.com 2024-11-22 08:41:41 a2ae0e30
│  foo
◆  z root() 00000000

🌐 vonz in /tmp/tmp.CIO7uhL2M8
at 08:41:44 ❯ echo bar > file

🌐 vonz in /tmp/tmp.CIO7uhL2M8
at 08:41:47 ❯ mv .jj/repo/store/git .git


🌐 vonz in /tmp/tmp.CIO7uhL2M8
at 08:42:15 ❯ echo -n '../../../.git' > .jj/repo/store/git_target


🌐 vonz in /tmp/tmp.CIO7uhL2M8
at 08:42:26 ❯ echo '/*' > .jj/.gitignore


🌐 vonz in /tmp/tmp.CIO7uhL2M8
at 08:42:30 ❯ git config --unset core.bare


🌐 vonz in /tmp/tmp.CIO7uhL2M8
at 08:42:34 ❯ jj new @-

Working copy now at: s b7c0c000 (empty) (no description set)
Parent commit      : u a2ae0e30 foo
Added 0 files, modified 1 files, removed 0 files

🌐 vonz in /tmp/tmp.CIO7uhL2M8
at 08:42:39 ❯ jj log
@  s martinvonz@google.com 2024-11-22 08:42:38 b7c0c000
│  (empty) (no description set)
│ ○  x martinvonz@google.com 2024-11-22 08:42:38 3f8453f7
├─╯  (no description set)
○  u martinvonz@google.com 2024-11-22 08:41:41 git_head() a2ae0e30
│  foo
◆  z root() 00000000
```

**Reply by nrbray** [reply]
Thank you [and thank you very much for jj].  

Sorry I should have read through the documentation and tried it out first.  Just starting out with jj I found that the step of ```new @-``` must have cleared my working copy (nothing was committed) and I should have discovered the undo command instead of trying commands I did not understand!

```sh
[nrb@dubedary:/tmp/tmp.a8T6Z3yy4a]$ jj git init; echo foo > file ; jj log
Initialized repo in "."
@  yupmlsst jj-mtd@email.to1.uk 2024-11-23 20:50:51 f046ec2a
│  (no description set)
◉  zzzzzzzz root() 00000000

[nrb@dubedary:/tmp/tmp.a8T6Z3yy4a]$ mv .jj/repo/store/git .git; echo -n '../../../.git' > .jj/repo/store/git_target; echo '/*' > .jj/.gitignore; git config --unset core.bare; jj log
@  yupmlsst jj-mtd@email.to1.uk 2024-11-23 20:50:51 f046ec2a
│  (no description set)
◉  zzzzzzzz root() 00000000

[nrb@dubedary:/tmp/tmp.a8T6Z3yy4a]$ jj new @-; jj log; cat file 
Working copy now at: uswzwwls 8f25022a (empty) (no description set)
Parent commit      : zzzzzzzz 00000000 (empty) (no description set)
Added 0 files, modified 0 files, removed 1 files
@  uswzwwls jj-mtd@email.to1.uk 2024-11-23 20:52:17 8f25022a
│  (empty) (no description set)
│ ◉  yupmlsst jj-mtd@email.to1.uk 2024-11-23 20:50:51 f046ec2a
├─╯  (no description set)
◉  zzzzzzzz root() 00000000
cat: file: No such file or directory

[nrb@dubedary:/tmp/tmp.a8T6Z3yy4a]$ jj undo
Working copy now at: yupmlsst f046ec2a (no description set)
Parent commit      : zzzzzzzz 00000000 (empty) (no description set)
Added 1 files, modified 0 files, removed 0 files

[nrb@dubedary:/tmp/tmp.a8T6Z3yy4a]$ ll; cat file 
total 4
-rw-r--r-- 1 nrb nrb 4 Nov 23 20:53 file
foo
```

**Comment by martinvonz** [comment]
> I could not for the life of me figure out how to restore my content. Not with jj, nor with my marginally better understanding of git!

If it's the problem Tim mentioned, then you simply need to update back to your old working-copy commit. You can find the commit in `jj log`. You should also be able to run `jj next` to do it without looking at the log.
